File: web/src/lib/zip.ts
//! Minimal ZIP archive (STORE method — no compression), read + write, no
//! dependency. Enough for bundling a set of small text files (CSV + manifest)
//! and reading them back. Only the STORE method is produced; on read, only
//! STORE (method 0) entries are supported — which is all this writer emits.

/** A single archive entry (text content). */
export type ZipEntry = { name: string; text: string };

const LOCAL_SIG = 0x04034b50;
const CENTRAL_SIG = 0x02014b50;
const EOCD_SIG = 0x06054b50;
const UTF8_FLAG = 0x0800; // filename is UTF-8 (general purpose bit 11)
const STORE = 0;
const DEFLATE = 8;

// CRC-32 (IEEE) lookup table, built once.
const CRC_TABLE = (() => {
  const t = new Uint32Array(256);
  for (let n = 0; n < 256; n++) {
    let c = n;
    for (let k = 0; k < 8; k++) c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1;
    t[n] = c >>> 0;
  }
  return t;
})();

function crc32(bytes: Uint8Array): number {
  let c = 0xffffffff;
  for (let i = 0; i < bytes.length; i++) c = CRC_TABLE[(c ^ bytes[i]) & 0xff] ^ (c >>> 8);
  return (c ^ 0xffffffff) >>> 0;
}

/** Builds a ZIP (STORE) archive from text entries and returns it as a Blob. */
export function zipStore(entries: ZipEntry[]): Blob {
  const enc = new TextEncoder();
  const files = entries.map((e) => ({ name: enc.encode(e.name), data: enc.encode(e.text) }));

  // Compute total size to allocate a single buffer.
  const localSize = (nameLen: number, dataLen: number) => 30 + nameLen + dataLen;
  const centralSize = (nameLen: number) => 46 + nameLen;
  let total = 22; // end of central directory
  for (const f of files) total += localSize(f.name.length, f.data.length) + centralSize(f.name.length);

  const buf = new Uint8Array(total);
  const view = new DataView(buf.buffer);
  let off = 0;
  const offsets: number[] = [];
  const crcs: number[] = [];

  // Local headers + data.
  for (const f of files) {
    offsets.push(off);
    const crc = crc32(f.data);
    crcs.push(crc);
    view.setUint32(off, LOCAL_SIG, true);
    view.setUint16(off + 4, 20, true); // version needed
    view.setUint16(off + 6, UTF8_FLAG, true); // flags
    view.setUint16(off + 8, 0, true); // method: store
    view.setUint16(off + 10, 0, true); // mod time
    view.setUint16(off + 12, 0, true); // mod date
    view.setUint32(off + 14, crc, true);
    view.setUint32(off + 18, f.data.length, true); // compressed size
    view.setUint32(off + 22, f.data.length, true); // uncompressed size
    view.setUint16(off + 26, f.name.length, true);
    view.setUint16(off + 28, 0, true); // extra len
    off += 30;
    buf.set(f.name, off);
    off += f.name.length;
    buf.set(f.data, off);
    off += f.data.length;
  }

  // Central directory.
  const centralStart = off;
  for (let i = 0; i < files.length; i++) {
    const f = files[i];
    view.setUint32(off, CENTRAL_SIG, true);
    view.setUint16(off + 4, 20, true); // version made by
    view.setUint16(off + 6, 20, true); // version needed
    view.setUint16(off + 8, UTF8_FLAG, true); // flags
    view.setUint16(off + 10, 0, true); // method
    view.setUint16(off + 12, 0, true); // mod time
    view.setUint16(off + 14, 0, true); // mod date
    view.setUint32(off + 16, crcs[i], true);
    view.setUint32(off + 20, f.data.length, true);
    view.setUint32(off + 24, f.data.length, true);
    view.setUint16(off + 28, f.name.length, true);
    view.setUint16(off + 30, 0, true); // extra len
    view.setUint16(off + 32, 0, true); // comment len
    view.setUint16(off + 34, 0, true); // disk number start
    view.setUint16(off + 36, 0, true); // internal attrs
    view.setUint32(off + 38, 0, true); // external attrs
    view.setUint32(off + 42, offsets[i], true); // local header offset
    off += 46;
    buf.set(f.name, off);
    off += f.name.length;
  }

  // End of central directory.
  view.setUint32(off, EOCD_SIG, true);
  view.setUint16(off + 4, 0, true); // disk number
  view.setUint16(off + 6, 0, true); // central dir disk
  view.setUint16(off + 8, files.length, true); // entries on this disk
  view.setUint16(off + 10, files.length, true); // total entries
  view.setUint32(off + 12, off - centralStart, true); // central dir size
  view.setUint32(off + 16, centralStart, true); // central dir offset
  view.setUint16(off + 20, 0, true); // comment len

  return new Blob([buf], { type: "application/zip" });
}

/** A located entry: where its bytes are and how they were stored. */
type RawEntry = { name: string; method: number; start: number; size: number };

/** Walks the central directory. The authority on what an archive contains — the
 * local headers repeat the same information and are allowed to lie about sizes
 * when an entry carries a trailing data descriptor. */
function listEntries(bytes: Uint8Array): RawEntry[] {
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const dec = new TextDecoder();

  // Locate the End Of Central Directory record (scan backwards; no comment
  // assumed, but tolerate one up to 64 KiB).
  let eocd = -1;
  for (let i = bytes.length - 22; i >= 0 && i >= bytes.length - 22 - 0xffff; i--) {
    if (view.getUint32(i, true) === EOCD_SIG) {
      eocd = i;
      break;
    }
  }
  if (eocd < 0) throw new Error("invalid zip: no EOCD");

  const count = view.getUint16(eocd + 10, true);
  let off = view.getUint32(eocd + 16, true); // central dir offset
  const out: RawEntry[] = [];

  for (let i = 0; i < count; i++) {
    if (view.getUint32(off, true) !== CENTRAL_SIG) throw new Error("invalid zip: bad central header");
    const method = view.getUint16(off + 10, true);
    const compSize = view.getUint32(off + 20, true);
    const nameLen = view.getUint16(off + 28, true);
    const extraLen = view.getUint16(off + 30, true);
    const commentLen = view.getUint16(off + 32, true);
    const localOff = view.getUint32(off + 42, true);
    const name = dec.decode(bytes.subarray(off + 46, off + 46 + nameLen));

    // Data starts after the local header (whose extra field length may differ).
    const localNameLen = view.getUint16(localOff + 26, true);
    const localExtraLen = view.getUint16(localOff + 28, true);
    const start = localOff + 30 + localNameLen + localExtraLen;
    out.push({ name, method, start, size: compSize });

    off += 46 + nameLen + extraLen + commentLen;
  }
  return out;
}

/** Reads a ZIP (STORE) archive into text entries. Non-STORE entries throw.
 * Used for bundles this app wrote itself, which are always STORE. */
export function unzip(bytes: Uint8Array): ZipEntry[] {
  const dec = new TextDecoder();
  return listEntries(bytes).map((e) => {
    if (e.method !== STORE) throw new Error(`unsupported zip method ${e.method} for ${e.name}`);
    return { name: e.name, text: dec.decode(bytes.subarray(e.start, e.start + e.size)) };
  });
}

/** Inflates a raw DEFLATE stream using the platform's own decompressor.
 *
 * No inflate implementation is bundled for this: `DecompressionStream` is in
 * every browser this app supports, and a compression codec is exactly the kind
 * of dependency worth not having. */
async function inflateRaw(data: Uint8Array): Promise<Uint8Array> {
  const stream = new Blob([data as BlobPart]).stream().pipeThrough(new DecompressionStream("deflate-raw"));
  return new Uint8Array(await new Response(stream).arrayBuffer());
}

/** Reads any ZIP this app is likely to be handed — STORE or DEFLATE — as raw
 * bytes, keyed by entry name.
 *
 * Bytes rather than text because an archive from elsewhere carries images
 * alongside its documents, and decoding a PNG as UTF-8 would quietly corrupt it.
 * Directory entries (trailing `/`) are dropped: the paths of the files say
 * everything about the tree. */
export async function unzipAll(bytes: Uint8Array): Promise<Map<string, Uint8Array>> {
  const out = new Map<string, Uint8Array>();
  for (const e of listEntries(bytes)) {
    if (e.name.endsWith("/")) continue;
    const raw = bytes.subarray(e.start, e.start + e.size);
    if (e.method === STORE) {
      out.set(e.name, raw);
    } else if (e.method === DEFLATE) {
      out.set(e.name, await inflateRaw(raw));
    } else {
      throw new Error(`unsupported zip method ${e.method} for ${e.name}`);
    }
  }
  return out;
}

File: src/backup/zip.rs
//! Minimal ZIP (STORE method only), reader and writer.
//!
//! Why hand-rolled rather than a crate: a backup archive holds a SQLite file and
//! a directory of already-compressed blobs, so there is nothing to gain from
//! deflate — and STORE means the archive is a container, not a codec. The whole
//! format we need is three fixed-layout records. `web/src/lib/zip.ts` makes the
//! same call on the frontend for page bundles; this is its counterpart.
//!
//! ZIP rather than tar because the person opening one may well be on Windows in
//! the middle of a bad day, and double-clicking has to work.
//!
//! Deliberately NOT supported: compression, encryption, Zip64. The first two are
//! not wanted; the third bounds an archive to 4 GiB, which `create` checks for
//! rather than silently truncating.

use std::io::{self, Read, Seek, SeekFrom, Write};
use std::sync::LazyLock;

const LOCAL_SIG: u32 = 0x0403_4b50;
const CENTRAL_SIG: u32 = 0x0201_4b50;
const EOCD_SIG: u32 = 0x0605_4b50;
/// General purpose bit 11: the name is UTF-8.
const UTF8_FLAG: u16 = 0x0800;
const STORE: u16 = 0;
/// 4 GiB — the point past which Zip64 becomes mandatory.
pub const MAX_ARCHIVE: u64 = u32::MAX as u64;

static CRC_TABLE: LazyLock<[u32; 256]> = LazyLock::new(|| {
    let mut t = [0u32; 256];
    for (n, slot) in t.iter_mut().enumerate() {
        let mut c = n as u32;
        for _ in 0..8 {
            c = if c & 1 != 0 { 0xedb8_8320 ^ (c >> 1) } else { c >> 1 };
        }
        *slot = c;
    }
    t
});

/// Running CRC-32 (IEEE), so an entry can be checksummed while it streams.
#[derive(Clone, Copy)]
pub struct Crc32(u32);

impl Default for Crc32 {
    fn default() -> Self {
        Self(0xffff_ffff)
    }
}

impl Crc32 {
    pub fn update(&mut self, bytes: &[u8]) {
        for b in bytes {
            self.0 = CRC_TABLE[((self.0 ^ u32::from(*b)) & 0xff) as usize] ^ (self.0 >> 8);
        }
    }
    pub fn finish(self) -> u32 {
        self.0 ^ 0xffff_ffff
    }
}

/// MS-DOS date and time, as ZIP stores them, from a Unix timestamp in seconds.
/// Out-of-range dates clamp to the epoch the format itself starts at (1980).
fn dos_datetime(unix_secs: i64) -> (u16, u16) {
    let Ok(dt) = time::OffsetDateTime::from_unix_timestamp(unix_secs) else {
        return (0x0021, 0); // 1980-01-01 00:00:00
    };
    let year = dt.year();
    if !(1980..=2107).contains(&year) {
        return (0x0021, 0);
    }
    let date = (((year - 1980) as u16) << 9) | ((u8::from(dt.month()) as u16) << 5) | dt.day() as u16;
    let time_ = ((dt.hour() as u16) << 11) | ((dt.minute() as u16) << 5) | (dt.second() as u16 / 2);
    (date, time_)
}

struct Central {
    name: String,
    crc: u32,
    size: u32,
    offset: u32,
    date: u16,
    time: u16,
}

/// Writes a STORE-only archive. Needs `Seek` because each local header carries
/// its entry's CRC and size, which are only known once the data has streamed
/// past — the header is written blank and patched. (The alternative, trailing
/// data descriptors, produces archives some tools read less happily.)
pub struct ZipWriter<W: Write + Seek> {
    out: W,
    entries: Vec<Central>,
}

impl<W: Write + Seek> ZipWriter<W> {
    pub fn new(out: W) -> Self {
        Self { out, entries: Vec::new() }
    }

    /// Streams one entry from `src`, checksumming as it goes. Never holds more
    /// than the copy buffer in memory, whatever the file's size.
    pub fn add<R: Read>(&mut self, name: &str, unix_secs: i64, src: &mut R) -> io::Result<()> {
        let (date, time_) = dos_datetime(unix_secs);
        let offset = self.out.stream_position()?;
        let name_bytes = name.as_bytes();

        // Local header, with CRC and sizes still unknown.
        self.out.write_all(&LOCAL_SIG.to_le_bytes())?;
        self.out.write_all(&20u16.to_le_bytes())?; // version needed
        self.out.write_all(&UTF8_FLAG.to_le_bytes())?;
        self.out.write_all(&STORE.to_le_bytes())?;
        self.out.write_all(&time_.to_le_bytes())?;
        self.out.write_all(&date.to_le_bytes())?;
        let placeholder = self.out.stream_position()?;
        self.out.write_all(&[0u8; 12])?; // crc + compressed + uncompressed
        self.out.write_all(&(name_bytes.len() as u16).to_le_bytes())?;
        self.out.write_all(&0u16.to_le_bytes())?; // extra field length
        self.out.write_all(name_bytes)?;

        let mut crc = Crc32::default();
        let mut size: u64 = 0;
        let mut buf = vec![0u8; 64 * 1024];
        loop {
            let n = src.read(&mut buf)?;
            if n == 0 {
                break;
            }
            crc.update(&buf[..n]);
            size += n as u64;
            self.out.write_all(&buf[..n])?;
        }
        let end = self.out.stream_position()?;
        if end > MAX_ARCHIVE {
            return Err(io::Error::other(
                "archive would exceed 4 GiB, which this writer does not support (no Zip64)",
            ));
        }
        let crc = crc.finish();

        // Back-fill the header now that the numbers are known.
        self.out.seek(SeekFrom::Start(placeholder))?;
        self.out.write_all(&crc.to_le_bytes())?;
        self.out.write_all(&(size as u32).to_le_bytes())?;
        self.out.write_all(&(size as u32).to_le_bytes())?;
        self.out.seek(SeekFrom::Start(end))?;

        self.entries.push(Central {
            name: name.to_string(),
            crc,
            size: size as u32,
            offset: offset as u32,
            date,
            time: time_,
        });
        Ok(())
    }

    pub fn add_bytes(&mut self, name: &str, unix_secs: i64, data: &[u8]) -> io::Result<()> {
        self.add(name, unix_secs, &mut io::Cursor::new(data))
    }

    /// Writes the central directory and the end-of-archive record.
    pub fn finish(mut self) -> io::Result<W> {
        let start = self.out.stream_position()?;
        for e in &self.entries {
            let name = e.name.as_bytes();
            self.out.write_all(&CENTRAL_SIG.to_le_bytes())?;
            self.out.write_all(&20u16.to_le_bytes())?; // version made by
            self.out.write_all(&20u16.to_le_bytes())?; // version needed
            self.out.write_all(&UTF8_FLAG.to_le_bytes())?;
            self.out.write_all(&STORE.to_le_bytes())?;
            self.out.write_all(&e.time.to_le_bytes())?;
            self.out.write_all(&e.date.to_le_bytes())?;
            self.out.write_all(&e.crc.to_le_bytes())?;
            self.out.write_all(&e.size.to_le_bytes())?;
            self.out.write_all(&e.size.to_le_bytes())?;
            self.out.write_all(&(name.len() as u16).to_le_bytes())?;
            self.out.write_all(&0u16.to_le_bytes())?; // extra
            self.out.write_all(&0u16.to_le_bytes())?; // comment
            self.out.write_all(&0u16.to_le_bytes())?; // disk number
            self.out.write_all(&0u16.to_le_bytes())?; // internal attrs
            self.out.write_all(&0u32.to_le_bytes())?; // external attrs
            self.out.write_all(&e.offset.to_le_bytes())?;
            self.out.write_all(name)?;
        }
        let end = self.out.stream_position()?;
        let count = self.entries.len() as u16;
        self.out.write_all(&EOCD_SIG.to_le_bytes())?;
        self.out.write_all(&0u16.to_le_bytes())?; // this disk
        self.out.write_all(&0u16.to_le_bytes())?; // disk with central dir
        self.out.write_all(&count.to_le_bytes())?;
        self.out.write_all(&count.to_le_bytes())?;
        self.out.write_all(&((end - start) as u32).to_le_bytes())?;
        self.out.write_all(&(start as u32).to_le_bytes())?;
        self.out.write_all(&0u16.to_le_bytes())?; // comment length
        self.out.flush()?;
        Ok(self.out)
    }
}

/// One entry, as listed by the central directory — which is the authority on
/// what an archive contains.
#[derive(Debug, Clone)]
pub struct Entry {
    pub name: String,
    pub size: u64,
    pub crc: u32,
    /// Offset of the entry's LOCAL header.
    pub offset: u64,
}

fn u16_at(b: &[u8], i: usize) -> u16 {
    u16::from_le_bytes([b[i], b[i + 1]])
}
fn u32_at(b: &[u8], i: usize) -> u32 {
    u32::from_le_bytes([b[i], b[i + 1], b[i + 2], b[i + 3]])
}

fn bad(msg: &str) -> io::Error {
    io::Error::new(io::ErrorKind::InvalidData, msg)
}

/// Lists the archive by reading its central directory.
///
/// Refuses anything it cannot honestly read — a compressed entry, a Zip64
/// archive, a truncated directory — rather than returning a partial listing that
/// a caller would mistake for the whole archive.
pub fn list<R: Read + Seek>(r: &mut R) -> io::Result<Vec<Entry>> {
    let len = r.seek(SeekFrom::End(0))?;
    if len < 22 {
        return Err(bad("not a ZIP archive: too short"));
    }
    // The EOCD sits at the end, possibly behind a comment of up to 64 KiB.
    let window = len.min(22 + 0xffff) as usize;
    r.seek(SeekFrom::End(-(window as i64)))?;
    let mut tail = vec![0u8; window];
    r.read_exact(&mut tail)?;
    let eocd = (0..=tail.len() - 22)
        .rev()
        .find(|&i| u32_at(&tail, i) == EOCD_SIG)
        .ok_or_else(|| bad("not a ZIP archive: no end-of-archive record"))?;

    let count = u16_at(&tail, eocd + 10) as usize;
    let dir_size = u32_at(&tail, eocd + 12) as usize;
    let dir_at = u32_at(&tail, eocd + 16) as u64;
    if dir_at == u32::MAX as u64 || count == 0xffff {
        return Err(bad("Zip64 archives are not supported"));
    }

    r.seek(SeekFrom::Start(dir_at))?;
    let mut dir = vec![0u8; dir_size];
    r.read_exact(&mut dir)?;

    let mut out = Vec::with_capacity(count);
    let mut at = 0usize;
    for _ in 0..count {
        if at + 46 > dir.len() || u32_at(&dir, at) != CENTRAL_SIG {
            return Err(bad("corrupt central directory"));
        }
        if u16_at(&dir, at + 10) != STORE {
            return Err(bad("compressed entries are not supported"));
        }
        let crc = u32_at(&dir, at + 16);
        let size = u32_at(&dir, at + 24) as u64;
        let name_len = u16_at(&dir, at + 28) as usize;
        let extra_len = u16_at(&dir, at + 30) as usize;
        let comment_len = u16_at(&dir, at + 32) as usize;
        let offset = u32_at(&dir, at + 42) as u64;
        let name_at = at + 46;
        if name_at + name_len > dir.len() {
            return Err(bad("corrupt central directory: truncated name"));
        }
        let name = String::from_utf8_lossy(&dir[name_at..name_at + name_len]).into_owned();
        out.push(Entry { name, size, crc, offset });
        at = name_at + name_len + extra_len + comment_len;
    }
    Ok(out)
}

/// Streams one entry's bytes into `w`, verifying its CRC before returning.
///
/// The check is the point: a backup archive is read exactly once, on the worst
/// day, and silently writing corrupt bytes over a live database would be the
/// single most destructive thing this codebase could do.
pub fn extract<R: Read + Seek, W: Write>(r: &mut R, e: &Entry, w: &mut W) -> io::Result<()> {
    r.seek(SeekFrom::Start(e.offset))?;
    let mut head = [0u8; 30];
    r.read_exact(&mut head)?;
    if u32_at(&head, 0) != LOCAL_SIG {
        return Err(bad("corrupt entry: bad local header"));
    }
    let name_len = u16_at(&head, 26) as u64;
    let extra_len = u16_at(&head, 28) as u64;
    r.seek(SeekFrom::Start(e.offset + 30 + name_len + extra_len))?;

    let mut crc = Crc32::default();
    let mut left = e.size;
    let mut buf = vec![0u8; 64 * 1024];
    while left > 0 {
        let want = buf.len().min(left as usize);
        r.read_exact(&mut buf[..want])?;
        crc.update(&buf[..want]);
        w.write_all(&buf[..want])?;
        left -= want as u64;
    }
    if crc.finish() != e.crc {
        return Err(bad("corrupt entry: checksum mismatch"));
    }
    w.flush()
}

/// Reads a whole entry into memory. For the small ones (the manifest) only.
pub fn read_all<R: Read + Seek>(r: &mut R, e: &Entry) -> io::Result<Vec<u8>> {
    let mut out = Vec::with_capacity(e.size as usize);
    extract(r, e, &mut out)?;
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Cursor;

    fn roundtrip(entries: &[(&str, Vec<u8>)]) -> Vec<u8> {
        let mut w = ZipWriter::new(Cursor::new(Vec::new()));
        for (name, data) in entries {
            w.add_bytes(name, 1_787_000_000, data).expect("add");
        }
        w.finish().expect("finish").into_inner()
    }

    #[test]
    fn writes_an_archive_it_can_read_back() {
        let big = vec![0xABu8; 200 * 1024]; // spans several copy buffers
        let bytes = roundtrip(&[
            ("backup.json", b"{\"format\":1}".to_vec()),
            ("bramblekeep.db", big.clone()),
            ("files/deadbeef", b"blob".to_vec()),
        ]);

        let mut r = Cursor::new(bytes);
        let entries = list(&mut r).expect("list");
        let names: Vec<_> = entries.iter().map(|e| e.name.as_str()).collect();
        assert_eq!(names, ["backup.json", "bramblekeep.db", "files/deadbeef"]);

        assert_eq!(read_all(&mut r, &entries[0]).unwrap(), b"{\"format\":1}");
        assert_eq!(read_all(&mut r, &entries[1]).unwrap(), big);
        assert_eq!(read_all(&mut r, &entries[2]).unwrap(), b"blob");
    }

    #[test]
    fn an_empty_entry_survives_the_round_trip() {
        let bytes = roundtrip(&[("empty", Vec::new())]);
        let mut r = Cursor::new(bytes);
        let entries = list(&mut r).expect("list");
        assert_eq!(entries[0].size, 0);
        assert!(read_all(&mut r, &entries[0]).unwrap().is_empty());
    }

    /// The guard that matters: a flipped byte must be caught, not written out.
    #[test]
    fn a_corrupted_entry_is_refused() {
        let mut bytes = roundtrip(&[("bramblekeep.db", vec![1u8; 5_000])]);
        // Flip a byte inside the entry's data, past every header.
        let at = bytes.len() / 2;
        bytes[at] ^= 0xff;

        let mut r = Cursor::new(bytes);
        let entries = list(&mut r).expect("list");
        let err = read_all(&mut r, &entries[0]).expect_err("checksum must fail");
        assert!(err.to_string().contains("checksum"), "got {err}");
    }

    #[test]
    fn refuses_something_that_is_not_an_archive() {
        let mut r = Cursor::new(b"this is not a zip file, not even close".to_vec());
        assert!(list(&mut r).is_err());
        let mut empty = Cursor::new(Vec::new());
        assert!(list(&mut empty).is_err());
    }

    #[test]
    fn crc32_matches_the_known_vector() {
        let mut c = Crc32::default();
        c.update(b"123456789");
        assert_eq!(c.finish(), 0xCBF4_3926); // the standard CRC-32/ISO-HDLC check value
    }
}
